core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -368,35 +392,50 @@ public void onFailure(Throwable e) {
 
         @Override
         protected void doRun() {
+            // shardID added to the request, as a side effect of calling resolveRequest
             if (checkBlocks() == false) {
                 return;
             }
-            final ShardIterator shardIt = shards(observer.observedState(), internalRequest);
-            final ShardRouting primary = resolvePrimary(shardIt);
-            if (primary == null) {
-                retryBecauseUnavailable(shardIt.shardId(), "No active shards.");
+            final IndexRoutingTable indexRoutingTable = observer.observedState().getRoutingTable().index(request.shardId().getIndex());
+            if (indexRoutingTable == null) {
+                logger.trace("index for shard [{}] is not found, scheduling a retry.", request.shardId());
+                retryBecauseUnavailable(request.shardId(), "index is not active");
                 return;
             }
-            if (primary.active() == false) {
-                logger.trace("primary shard [{}] is not yet active, scheduling a retry.", primary.shardId());
-                retryBecauseUnavailable(shardIt.shardId(), "Primary shard is not active or isn't assigned to a known node.");
+            final IndexShardRoutingTable shardRoutingTable = indexRoutingTable.shard(request.shardId().getId());
+            if (shardRoutingTable == null) {
+                logger.trace("routing for shard [{}] is not found, scheduling a retry.", request.shardId());
+                retryBecauseUnavailable(request.shardId(), "primary shard is not active");
```

Comment:
same comment. We shouldn't retry in this case. [Old code throws ShardNotFoundException](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/cluster/routing/OperationRouting.java#L230)